man/man8/xfs_healer_start.8
.TH xfs_healer_start 8
.SH NAME
xfs_healer_start \- starts xfs_healer instances
.SH SYNOPSIS
.B xfs_healer_start
[
.B OPTIONS
]
.br
.B xfs_healer \-V
.SH DESCRIPTION
.B xfs_healer_start
starts the xfs_healer service whenever the kernel mounts an XFS filesystem in
the current mount namespace.
.PP
.B WARNING!
This program is
.BR EXPERIMENTAL ","
which means that its behavior and interface
could change at any time!

Normally this program runs as a systemd service.

.SH OPTIONS
.TP
.B \-\-supported
Check if the kernel supports listening for mount events.
Exits with 0 if it does, and non-zero if not.
.TP
.BI "\-\-mountns " path
Monitor the given mount namespace.
Defaults to the mount namespace associated with the process itself.
.TP
.BI \-V
Prints the version number and exit.
.SH SEE ALSO
.BR xfs_healer (8).


debian/changelog
xfsprogs (6.1.0-rc0~WIP-2022-10-14-1djwong1) jammy-experimental; urgency=low

  * Base off for-next on 2022-08-30 (5.19.0+)
  * Support bmbt repairs in xfs_repair.
  * Support realtime rmap btree.
  * Metadata directories.
  * Support atomic file range swaps.
  * AG status and noalloc.
  * Realtime reflink.
  * Better balanced repair workers for scrub.
  * Scrub repair dependencies.
  * reflink/rmapbt/finobt/metadir upgrades.
  * Vectorized scrub experiment.
  * Support for cowextsize on rtreflink files.
  * GETFSREFCOUNTS ioctl and xfs_io utility.
  * Static checking is clean enough (2021-07-15).
  * Prototype free space defragmentation.
  * In-memory reverse mapping btrees for repair.
  * Atomic symlink swapext.
  * Sample mkfs configuration files.
  * Faster xfs_scrub phase5.
  * Scrub file data once per month without a bunch of competing timers.
  * Larger minimum log size and drop support for small filesystems.
  * Multiarch support for libhandle.
  * Better security for xfs_scrub background services.
  * Improve xfs_scrub_all resiliency when systemd gets re-exec'd.
  * Realtime allocation groups.
  * Fixed CPU usage resource control for xfs_scrub processes.
  * Intent items take perag/rtgroup references.

 -- Darrick J. Wong <djwong@djwong.org>  Fri, 14 Oct 2022 09:58:00 -0700

xfsprogs (5.19.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 12 Aug 2022 13:40:25 -0500

xfsprogs (5.18.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 03 Jun 2022 15:18:30 -0400

xfsprogs (5.16.0-1) unstable; urgency=medium

  * New upstream release

  [ Bastian Germann ]
  * debian: Generate .gitcensus instead of .census (Closes: #999743)

 -- Nathan Scott <nathans@debian.org>  Wed, 04 May 2022 14:50:48 -0400

xfsprogs (5.15.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Wed, 06 Apr 2022 16:31:00 -0400

xfsprogs (5.14.2-1) unstable; urgency=medium

  * New upstream release
  * Move rogue fallthrough macro out of linux.h (Closes: #1000974)

 -- Nathan Scott <nathans@debian.org>  Mon, 06 Dec 2021 14:26:57 -0500

xfsprogs (5.14.1-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Thu, 02 Dec 2021 15:26:38 -0500

xfsprogs (5.14.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 19 Nov 2021 10:51:02 -0500

xfsprogs (5.14.0-rc1-1) unstable; urgency=medium

  [ Dave Chinner ]
  * New build dependency: liburcu-dev

  [ Helmut Grohne ]
  * Fix FTCBFS (Closes: #794158)
    + Pass --build and --host to configure

  [ Boian Bonev ]
  * Fix FTBFS (Closes: #997656)
    + Keep custom install-sh after autogen

 -- Bastian Germann <bage@debian.org>  Sun, 14 Nov 2021 23:18:22 +0100

xfsprogs (5.13.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 20 Aug 2021 12:03:31 -0400

xfsprogs (5.12.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 21 May 2021 15:52:00 -0400

xfsprogs (5.11.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 12 Mar 2021 14:08:29 -0500

xfsprogs (5.11.0-rc0-1) experimental; urgency=medium

  [ Steve Langasek ]
  * Regenerate config.guess using debhelper

  [ Bastian Germann ]
  * Build-depend on libinih-dev with udeb package

 -- Bastian Germann <bastiangermann@fishpost.de>  Sat, 20 Feb 2021 11:57:31 +0100

xfsprogs (5.10.0-3) unstable; urgency=medium

  * Drop unused dh-python from Build-Depends (Closes: #981361)
  * Only build for Linux
  * Prevent installing duplicate changelog (Closes: #570704)

 -- Bastian Germann <bastiangermann@fishpost.de>  Fri, 05 Feb 2021 00:18:31 +0100

xfsprogs (5.10.0-2) unstable; urgency=low

  * Team upload
  * debian: cryptographically verify upstream tarball (Closes: #979644)
  * debian: remove dependency on essential util-linux
  * debian: remove "Priority: extra"
  * debian: use Package-Type over its predecessor
  * debian: add missing copyright info (Closes: #979653)

 -- Bastian Germann <bastiangermann@fishpost.de>  Thu, 14 Jan 2021 18:59:14 +0100

xfsprogs (5.10.0-1) unstable; urgency=low

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 11 Dec 2020 17:16:52 -0500

xfsprogs (5.10.0-rc1-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Fri, 04 Dec 2020 12:19:10 -0500

xfsprogs (5.10.0-rc0-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Mon, 16 Nov 2020 14:26:14 -0600

xfsprogs (5.9.0-1) unstable; urgency=low

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Tue, 20 Oct 2020 11:38:49 -0400

xfsprogs (5.9.0-rc1-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Tue, 13 Oct 2020 12:18:43 -0400

xfsprogs (5.9.0-rc0-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Thu, 17 Sep 2020 09:11:13 -0500

xfsprogs (5.8.0-1) unstable; urgency=low

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 04 Sep 2020 15:40:01 -0400

xfsprogs (5.8.0-rc1-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Wed, 26 Aug 2020 14:06:48 -0500

xfsprogs (5.8.0-rc0-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Mon, 10 Aug 2020 14:42:12 -0500

xfsprogs (5.7.0-1) unstable; urgency=low

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 24 Jul 2020 11:30:00 -0700

xfsprogs (5.7.0-rc1-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Wed, 15 Jul 2020 13:05:24 -0700

xfsprogs (5.7.0-rc0-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Thu, 07 May 2020 13:06:26 -0500


healer/xfs_healer_start.c
// SPDX-License-Identifier: GPL-2.0-or-later
/*
 * Copyright (C) 2026 Oracle.  All Rights Reserved.
 * Author: Darrick J. Wong <djwong@kernel.org>
 */
#include "xfs.h"

#include <errno.h>
#include <err.h>
#include <stdlib.h>
#include <stdio.h>
#include <fcntl.h>
#include <sys/fanotify.h>
#include <sys/types.h>
#include <unistd.h>
#include <linux/mount.h>
#include <sys/syscall.h>
#include <string.h>
#include <limits.h>

#include "platform_defs.h"
#include "libfrog/systemd.h"

#define DEFAULT_MOUNTNS_FILE		"/proc/self/ns/mnt"

static int debug = 0;
static const char *progname = "xfs_healer_start";

/* Start the xfs_healer service for a given mountpoint. */
static void
start_healer(
	const char	*mntpoint)
{
	char		unitname[PATH_MAX];
	int		ret;

	ret = systemd_path_instance_unit_name(XFS_HEALER_SVCNAME, mntpoint,
			unitname, PATH_MAX);
	if (ret) {
		fprintf(stderr, "%s: %s\n", mntpoint,
				_("Could not determine xfs_healer unit name."));
		return;
	}

	/*
	 * Restart so that we aren't foiled by an existing unit that's slowly
	 * working its way off a cycled mount.
	 */
	ret = systemd_manage_unit(UM_RESTART, unitname);
	if (ret) {
		fprintf(stderr, "%s: %s: %s\n", mntpoint,
				_("Could not start xfs_healer service unit"),
				unitname);
		return;
	}

	printf("%s: %s\n", mntpoint, _("xfs_healer service started."));
	fflush(stdout);
}

#define REQUIRED_STATMOUNT_FIELDS (STATMOUNT_FS_TYPE | \
				   STATMOUNT_MNT_POINT | \
				   STATMOUNT_MNT_ROOT)

/* Process a newly discovered mountpoint. */
static void
examine_mount(
	int			mnt_ns_fd,
	uint64_t		mnt_id)
{
	struct mnt_id_req	req = {
		.size		= sizeof(req),
		.mnt_id		= mnt_id,
#ifdef HAVE_LISTMOUNT_NS_FD
		.mnt_ns_fd	= mnt_ns_fd,
#else
		.spare		= mnt_ns_fd,
#endif
		.param		= REQUIRED_STATMOUNT_FIELDS,
	};
	size_t			smbuf_size = sizeof(struct statmount) + 4096;
	struct statmount	*smbuf = alloca(smbuf_size);
	int			ret;

	ret = syscall(SYS_statmount, &req, smbuf, smbuf_size, 0);
	if (ret) {
		perror("statmount");
		return;
	}

	if (debug) {
		printf("mount: id 0x%llx fstype %s mountpoint %s mntroot %s\n",
				(unsigned long long)mnt_id,
				(smbuf->mask & STATMOUNT_FS_TYPE) ?
					smbuf->str + smbuf->fs_type : "null",
				(smbuf->mask & STATMOUNT_MNT_POINT) ?
					smbuf->str + smbuf->mnt_point : "null",
				(smbuf->mask & STATMOUNT_MNT_ROOT) ?
					smbuf->str + smbuf->mnt_root : "null");
		fflush(stdout);
	}

	/* Look for mount points for the root dir of an XFS filesystem. */
	if ((smbuf->mask & REQUIRED_STATMOUNT_FIELDS) !=
			   REQUIRED_STATMOUNT_FIELDS)
		return;

	if (!strcmp(smbuf->str + smbuf->fs_type, "xfs") &&
	    !strcmp(smbuf->str + smbuf->mnt_root, "/"))
		start_healer(smbuf->str + smbuf->mnt_point);
}

/* Translate fanotify mount events into something we can process. */
static void
handle_mount_event(
	const struct fanotify_event_metadata	*event,
	int					mnt_ns_fd)
{
	const struct fanotify_event_info_header	*info;
	const struct fanotify_event_info_mnt	*mnt;
	int					off;

	if (event->fd != FAN_NOFD) {
		if (debug)
			fprintf(stderr, "Expected FAN_NOFD, got fd=%d\n",
					event->fd);
		return;
	}

	switch (event->mask) {
	case FAN_MNT_ATTACH:
		if (debug) {
			printf("FAN_MNT_ATTACH (len=%d)\n", event->event_len);
			fflush(stdout);
		}
		break;
	default:
		/* should never get here */
		return;
	}

	for (off = sizeof(*event) ; off < event->event_len;
	     off += info->len) {
		info = (struct fanotify_event_info_header *)
			((char *) event + off);

		switch (info->info_type) {
		case FAN_EVENT_INFO_TYPE_MNT:
			mnt = (struct fanotify_event_info_mnt *) info;

			if (debug) {
				printf( "Mount record: len=%d mnt_id=0x%llx\n",
						mnt->hdr.len, mnt->mnt_id);
				fflush(stdout);
			}

			examine_mount(mnt_ns_fd, mnt->mnt_id);
			break;

		default:
			if (debug)
				fprintf(stderr,
 "Unexpected fanotify event info_type=%d len=%d\n",
						info->info_type, info->len);
			break;
		}
	}
}

/* Extract mount attachment notifications from fanotify. */
static void
handle_notifications(
	char				*buffer,
	ssize_t				len,
	int				mnt_ns_fd)
{
	struct fanotify_event_metadata	*event =
		(struct fanotify_event_metadata *) buffer;

	for (; FAN_EVENT_OK(event, len); event = FAN_EVENT_NEXT(event, len)) {

		switch (event->mask) {
		case FAN_MNT_ATTACH:
			handle_mount_event(event, mnt_ns_fd);
			break;
		default:
			if (debug)
				fprintf(stderr,
 "Unexpected fanotify mark: 0x%llx\n",
					(unsigned long long)event->mask);
			break;
		}
	}
}

/* Start healer services for existing XFS mounts. */
static int
start_existing_mounts(
	int			mnt_ns_fd)
{
	struct mnt_id_req	req = {
		.size		= sizeof(struct mnt_id_req),
#ifdef HAVE_LISTMOUNT_NS_FD
		.mnt_ns_fd	= mnt_ns_fd,
#else
		.spare		= mnt_ns_fd,
#endif
		.mnt_id		= LSMT_ROOT,
	};
	uint64_t		mnt_ids[32];
	int			i;
	int			ret;

	while ((ret = syscall(SYS_listmount, &req, &mnt_ids, 32, 0)) > 0) {
		for (i = 0; i < ret; i++)
			examine_mount(mnt_ns_fd, mnt_ids[i]);

		req.param = mnt_ids[ret - 1];
	}

	if (ret < 0) {
		if (errno == ENOSYS)
			fprintf(stderr, "%s\n",
 _("This program requires the listmount system call."));
		else
			perror("listmount");
		return -1;
	}

	return 0;
}

static void __attribute__((noreturn))
usage(void)
{
	fprintf(stderr, "%s %s %s\n", _("Usage:"), progname, _("[OPTIONS]"));
	fprintf(stderr, "\n");
	fprintf(stderr, _("Options:\n"));
	fprintf(stderr, _("  --debug      Enable debugging messages.\n"));
	fprintf(stderr, _("  --mountns    Path to the mount namespace file.\n"));
	fprintf(stderr, _("  --supported  Make sure we can actually run.\n"));
	fprintf(stderr, _("  -V           Print version.\n"));

	exit(EXIT_FAILURE);
}

enum long_opt_nr {
	LOPT_DEBUG,
	LOPT_HELP,
	LOPT_MOUNTNS,
	LOPT_SUPPORTED,

	LOPT_MAX,
};

int
main(
	int		argc,
	char		*argv[])
{
	char		buffer[BUFSIZ];
	const char	*mntns = NULL;
	int		mnt_ns_fd;
	int		fan_fd;
	int		c;
	int		option_index;
	int		support_check = 0;
	int		ret = 0;

	struct option long_options[] = {
		[LOPT_SUPPORTED] = {"supported", no_argument, &support_check, 1 },
		[LOPT_DEBUG]	 = {"debug", no_argument, &debug, 1 },
		[LOPT_HELP]	 = {"help", no_argument, NULL, 0 },
		[LOPT_MOUNTNS]	 = {"mountns", required_argument, NULL, 0 },
		[LOPT_MAX]	 = {NULL, 0, NULL, 0 },
	};

	while ((c = getopt_long(argc, argv, "V", long_options, &option_index))
			!= EOF) {
		switch (c) {
		case 0:
			switch (option_index) {
			case LOPT_MOUNTNS:
				mntns = optarg;
				break;
			case LOPT_HELP:
				usage();
				break;
			default:
				break;
			}
			break;
		case 'V':
			fprintf(stdout, "%s %s %s\n", progname, _("version"),
					VERSION);
			fflush(stdout);
			return EXIT_SUCCESS;
		default:
			usage();
			break;
		}
	}

	/*
	 * Try to open the mount namespace file for the current process.
	 * fanotify requires this mount namespace file to send mount attachment
	 * events, so this is required for correct functionality.
	 */
	mnt_ns_fd = open(mntns ? mntns : DEFAULT_MOUNTNS_FILE, O_RDONLY);
	if (mnt_ns_fd < 0) {
		if (errno == ENOENT && !mntns) {
			perror(DEFAULT_MOUNTNS_FILE);
			fprintf(stderr, "%s\n",
 _("This program requires mount namespace support."));
		} else {
			perror(mntns ? mntns : DEFAULT_MOUNTNS_FILE);
		}
		ret = 1;
		goto out;
	}

	fan_fd = fanotify_init(FAN_REPORT_MNT, O_RDONLY);
	if (fan_fd < 0) {
		perror("fanotify_init");
		if (errno == EINVAL)
			fprintf(stderr, "%s\n",
 _("This program requires fanotify mount event support."));
		ret = 1;
		goto out;
	}

	ret = fanotify_mark(fan_fd, FAN_MARK_ADD | FAN_MARK_MNTNS,
			FAN_MNT_ATTACH, mnt_ns_fd, NULL);
	if (ret) {
		perror("fanotify_mark");
		goto out;
	}

	if (support_check) {
		/*
		 * We're being run as an ExecCondition process and we've
		 * decided to start the main service.  There is no need to wait
		 * for journald because the ExecStart version of ourselves will
		 * take care of the waiting for us.
		 */
		return systemd_service_exit_now(0);
	}

	if (debug) {
		printf("fanotify active\n");
		fflush(stdout);
	}

	ret = start_existing_mounts(mnt_ns_fd);
	if (ret)
		goto out;

	while (1) {
		ssize_t bytes_read = read(fan_fd, buffer, BUFSIZ);

		if (bytes_read < 0) {
			perror("fanotify");
			ret = 1;
			break;
		}

		handle_notifications(buffer, bytes_read, mnt_ns_fd);
	}

out:
	return systemd_service_exit(ret);
}
